File: Tests/Python_Sandbox/Pipeline/Utilities/CBASS_U_GetTStampEventIndex.py
import numpy as np

def GetTStampEventIndex(db1TStamp, db1EventTStamp, sOPTION):
    '''
    IN1EVENTIDX = NS_GetTStampEventIndex(DB1TSTAMP, DB1EVENTTSTAMP)

    Finds the indices of the  continues and monotonically increasing timestamp
    vector DB1TSTAMP that are closest to the event timestamps provided in
    DB1EVENTTSTAMP. Returns NaN for any event timestamps situated outside the
    range of values of DB1TSTAMP.

    '''

    # Checks if the values of db1TStamp are monotonically increasing
    verbose = sOPTION.blVerbose
    db1TStamp = db1TStamp.flatten()
    if np.any(np.diff(db1TStamp) <= 0):
        print('The values of db1TStamps should be monotonically increasing. Rescue mode enabled.')
        blRescueMode = True
        
    else: blRescueMode = False   

    # Sorts the events by timestamps
    in1SortIdx   = np.argsort(db1EventTStamp)
    # if verbose: print('in1SortIdx: ',in1SortIdx)
    db1EventTStamp = db1EventTStamp[in1SortIdx]
    # if verbose: print('db1EventTStamp: ',db1EventTStamp)

    # Gets the interval in the time stamp trace
    # db1TSInterval = np.array([np.diff(db1TStamp)/2, np.median(np.diff(db1TStamp)/2)])[0]
    dbTSInterval = np.median(np.diff(db1TStamp)/2)
    # if verbose: print('dbTSInterval: ',dbTSInterval)

    # Finds the closest index to each 
    in1EventIdx = np.empty_like(db1EventTStamp, dtype=np.int32)
    # if verbose: print('in1EventIdx.shape: ',in1EventIdx.shape)
    TSIdx = 0
    TStart = np.sum(db1EventTStamp < db1TStamp[0] - dbTSInterval) # Starts with the first index that is within the trace
    # if verbose: print('TStart: ',TStart)

    for jj in range(TStart,len(db1EventTStamp)):
        blMatch = False
        while blMatch == False and TSIdx <= len(db1TStamp):
            # if db1EventTStamp[jj] <= db1TStamp[TSIdx] + db1TSInterval:
            if np.abs(db1TStamp[TSIdx] - db1EventTStamp[jj]) <= dbTSInterval:
                blMatch = True
                in1EventIdx[jj] = TSIdx
                # if verbose: print('in1EventIdx[{}]: {}'.format(jj,in1EventIdx[jj]))
            else:
                TSIdx = TSIdx + 1
    
    return in1EventIdx

File: Tests/Matlab_Sandbox/CBASS_U_GetTStampEventIndex.m
function [in1EventIdx] = CBASS_U_GetTStampEventIndex(db1TStamp, db1EventTStamp)
%IN1EVENTIDX = CBASS_U_GetTStampEventIndex(DB1TSTAMP, DB1EVENTTSTAMP)
%
%Finds the indices of the  continues and monotonically increasing timestamp
%vector DB1TSTAMP that are closest to the event timestamps provided in
%DB1EVENTTSTAMP. Returns NaN for any event timestamps situated outside the
%range of values of DB1TSTAMP.
%
%2016-10-12 QP: Created

%Checks if the values of db1TStamp are monotonically increasing
if ~isvector(db1TStamp)
    error('db1TStamps should be a vector\r')
end
db1TStamp = db1TStamp(:);
if any(diff(db1TStamp) <= 0)
    error('The values of db1TStamps should be monotonically increasing\r')
end
% if length(unique(diff(db1TStamp))) ~= 1
%     error('The interval of db1TStamps is not constant')
% end
 
%Sorts the events by timestamps
[db1EventTStamp, in1SortIdx]   = sort(db1EventTStamp);
[~, in1UnsortIdx]              = sort(in1SortIdx);

%Gets the interval in the time stamp trace
dbTSInterval = median(diff(db1TStamp))/2;

%Finds the closest index to each 
in1EventIdx = nan(1, length(db1EventTStamp));
TSIdx = 1;
TStart = 1 + sum(db1EventTStamp < db1TStamp(1) - dbTSInterval); %Starts with the first index that is within the trace

for jj = TStart:length(db1EventTStamp)
    blMatch = false;
    while blMatch == false && TSIdx <= length(db1TStamp)
        if abs(db1TStamp(TSIdx) - db1EventTStamp(jj)) <= dbTSInterval
            blMatch = true;
            in1EventIdx(jj) = TSIdx;
        else
            TSIdx = TSIdx + 1;
        end
    end
end

%Unsort the timestamps
in1EventIdx = in1EventIdx(in1UnsortIdx);